### DISCUSSION TITLE
How to set Timezone for each user

### DISCUSSION THREAD

**Discussion by troycarpenter** [ampache/ampache#3590] [Q&A]
This scope of this question morphed from setting the timezone in the docker container, to now having a timezone setting for each user...below is the original discussion.  The feature request can be found [here](https://github.com/ampache/ampache/issues/3592)

I've noticed that all my logs, notifications, and anything that generally uses a timestamp is all in UTC.  Has anyone figured out how to get the system to report the times appropriate for the local timezone?

I have tried using the TZ=America/Chicago variable in the docker-compose file, but even though the system responds with the correct timezone when I type "date" on the cli, it doesn't change the logs/notification timestamps.

While it would be nice if this were a per-user setting since I have some users in other timezones, I would at least want timestamps to be in the timezone local to the server.

**Comment by troycarpenter** [comment]
Nobody?  I'm the only one that finds it odd that there are number of users listening to things at 2am? 

I haven't even been able to adjust the timestamps in real time, so I'm still stuck.

**Comment by troycarpenter** [comment]
After some investigation, the biggest bang for the buck is to modify **src/Config/functions.php**, in the **get_datetime** function.  Actually, I found that this is already handled IF the timezone is somehow set correctly.  

On line 450, there's a function call:
```
$format = new IntlDateFormatter($locale, $date_type, $time_type, date_default_timezone_get(), null, $pattern);
```
It turns out that **date_defualt_timezone_get()** is supposed to return the timezone as set in php.ini.  For me it isn't working and is always returning UTC.  However, if the timezone is something else, then that function will calculate and apply the timezone offset to the timestamp that is being generated.  By this I'm assuming that Ampache is storing all events in UTC and NOT the local timezone.

To make this work on a per user basis, it would simply be passing the user's  timezone (i.e. America/New_York) to the  **IntlDateFormatter()** function above instead of the system's timezone .  The question is where this timezone would come from.  Can it be captured by the server from the user's browser, or should the user put it in their server preference page (like under the part where they select their language).  I would suggest that is the best solution and defaulting to UTC when the user is created.

Once the preference variable is in place, then fetching that string at the start of the **get_datetime()** function and using it would then display items like timelines in the user's local timezone.  Then a user can tell if their friend in another timezone listened to a song at the same time as they did by looking at the other user's timeline.

I don't know if that breaks other non-timezone dependent output, or if there are more places that would need to be touched.  However, the get_datetime() routine does appear to be used in a lot of places to display timestamps for events.

I have tested by statically adding a timezone string (**$utz="America/New_York"**), using and using it like this on line 450:
```
$format = new IntlDateFormatter($locale, $date_type, $time_type, $utz, null, $pattern);
```
That alone caused all the displayed timestamps to shift to that timezone.  Other places I saw where the timestamp changed included the user timestamps for "member since" and "last seen".  I'm sure there are other subtle places I've not noticed yet.

**Reply by lachlan-00** [reply]
it looks like date_default_timezone_get() is null on docker so we should look at adding a config/option in the database as a fallback

I'll make it a server config option using the php timezone list
